feat: added Optimism deposited transaction support

merklefruit opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests.

Contributors guide: https://github.com/gakonst/ethers-rs/blob/master/CONTRIBUTING.md

The contributors guide includes instructions for running rustfmt and building the
documentation.
-->

## Motivation

Optimism has an extra transaction type: the Deposited transaction (type 0x7E), with 3 new transaction fields as described [in the spec here](https://github.com/ethereum-optimism/optimism/blob/develop/specs/deposits.md#the-deposited-transaction-type). The RLP encoding order is also different.

<!--
Explain the context and why you're making that change. What is the problem
you're trying to solve? In some cases there is not a problem and this can be
thought of as being the motivation for your change.
-->

## Solution

This PR introduces a new feature "optimism" much like the existing "celo" one, which enables the 3 new optional transaction fields.

<!--
Summarize the solution and provide any necessary context needed to understand
the code change.
-->

## PR Checklist

-   [x] Added Tests
-   [x] Added Documentation

prestwich commented:
I'm supportive of adding first-class optimism objects. However, I am concerned that this change might be too large for our current release cycle. We're trying to put this version into maintenance so that we can focus on ethers@3. 

In addition, our opinion generally is that the Celo feature was a bad approach to network-specific behavior in the first place, and this ought to be solved by abstraction of the argument and return types in the Middleware trait

@gakonst do you think it is worth it to unblock optimism folks?

gakonst reviewed (approved):
I would like to add this because it also unblocks op-reth work for client side. I am open to merging this if CI passes

prestwich commented on ethers-core/src/types/transaction/response.rs:
```diff
@@ -65,6 +65,22 @@ pub struct Transaction {
+    ///////////////// Optimism-specific transaction fields //////////////
+    /// The source-hash that uniquely identifies the origin of the deposit
+    #[cfg(feature = "optimism")]
+    #[serde(skip_serializing_if = "Option::is_none", rename = "sourceHash")]
```
needs `#[serde(default)]`

prestwich commented on ethers-core/src/types/transaction/response.rs:
```diff
@@ -65,6 +65,22 @@ pub struct Transaction {
+
+    /// The ETH value to mint on L2
+    #[cfg(feature = "optimism")]
+    #[serde(skip_serializing_if = "Option::is_none")]
```
needs `#[serde(default)]`

prestwich commented on ethers-core/src/types/transaction/response.rs:
```diff
@@ -65,6 +65,22 @@ pub struct Transaction {
+
+    /// True if the transaction does not interact with the L2 block gas pool
+    #[cfg(feature = "optimism")]
+    #[serde(skip_serializing_if = "Option::is_none", rename = "isSystemTx")]
```
needs `#[serde(default)]`

prestwich commented on ethers-core/src/types/transaction/response.rs:
```diff
@@ -173,6 +191,20 @@ impl Transaction {
+            // Optimism Deposited Transaction
+            #[cfg(feature = "optimism")]
+            Some(x) if x == U64::from(0x7E) => {
+                rlp.append(&self.source_hash.expect("sourceHash is required for deposited tx"));
```
should `source_hash` be `<_>` instead of `Option<_>`

prestwich replied:
oh i thought we were in a TypedTransaction context. I suppose this is fine here, but you should use `rlp_opt`, not a panic

you'll also need to add a TypedTransaction enum variant

merklefruit replied:
> you'll also need to add a TypedTransaction enum variant

Ok, @prestwich regarding this do you think we should reuse the `Eip1559TransactionRequest` type and extend it with the optimism feature, or instead create a new `OptimismDepositedTransactionRequest` type?

prestwich replied:
if optimism supports both separately, do both

prestwich reviewed (changes requested):
Need to update CI to run with the optimsim feature enabled. 
Need to document that Optimism and Celo are mutually exclusive

DaniPopes commented on README.md:
````diff
@@ -58,6 +58,22 @@ You can get one [here](https://polygonscan.io/apis).
+
+```toml
+[dependencies]
+ethers = { version = "1.0.2", features = ["optimism"] }
````
```suggestion
ethers = { version = "2.0", features = ["optimism"] }
```

merklefruit commented:
Ok, I've split the addition of a new TypedTransaction variant to a secondary PR to make the diffs cleaner. 
Here: https://github.com/merklefruit/ethers-rs/pull/1

I've also added a couple tests for RLP encoding/decoding of this new tx type. Let me know what other kinds of tests I should eventually add!

merklefruit commented:
@gakonst I believe the PR above (https://github.com/merklefruit/ethers-rs/pull/1) was not merged into this branch yet. I've closed it now, so I need to reopen a PR from this branch
